### ISSUE TITLE
Wakatime just stopped working

### ISSUE THREAD

**Issue by andrefangeloni** [wakatime/wakatime-cli#770]
**Expected behavior (what should have happened)**:
Should to track activity

**Actual behavior (what went wrong)**:
My analytics are not syncing

**Environment**:

- OS:  macOS Monterey (12.5.1)
- Platform:  Intel
- IDE: VS Code

**Logs**:
```json
{"caller":"cmd/run.go:270","func":"cmd.runCmd","level":"error","message":"failed to run command: today fetch failed: failed fetching today from api: failed to make request to \"https://api.wakatime.com/api/v1/users/current/statusbar/today\": Get \"https://api.wakatime.com/api/v1/users/current/statusbar/today\": EOF","now":"2022-08-26T21:48:51-03:00","version":"v1.54.0"}
{"caller":"cmd/run.go:270","func":"cmd.runCmd","level":"error","message":"failed to run command: sending heartbeat(s) later because currently rate limited: won't send heartbeat due to backoff","now":"2022-08-26T22:00:24-03:00","version":"v1.54.0"}
```

Wakatime always worked well on my machine, but today I saw a message on status bar wakatimeError. I generated a new API Key, disabled and uninstalled extension, installed again and now the timer always show 0 sec.

What I need to do?
Thanks in advance

**Comment by alanhamlett** [comment]
That error is a network error when updating the status bar. Are there any or errors in your log related to sending heartbeats?

**Comment by andrefangeloni** [comment]
> That error is a network error when updating the status bar. Are there any or errors in your log related to sending heartbeats?

Hi @alanhamlett,
I started my machine right now and these are the logs are showing up to me:

```json
{"caller":"cmd/run.go:270","func":"cmd.runCmd","level":"error","message":"failed to run command: today fetch failed: failed fetching today from api: failed to make request to \"https://api.wakatime.com/api/v1/users/current/statusbar/today\": Get \"https://api.wakatime.com/api/v1/users/current/statusbar/today\": EOF","now":"2022-08-27T00:12:52-03:00","version":"v1.54.0"}
{"caller":"pkg/api/api.go:64","func":"api.(*Client)","level":"warning","message":"dns error: https://api.wakatime.com/api/v1/users/current/heartbeats.bulk. Retrying with fallback dns resolver","now":"2022-08-27T10:39:42-03:00","version":"v1.54.0"}
{"caller":"cmd/run.go:270","func":"cmd.runCmd","level":"error","message":"failed to run command: offline sync failed: failed to sync offline heartbeats: failed making request to \"https://api.wakatime.com/api/v1/users/current/heartbeats.bulk\": Post \"https://api.wakatime.com/api/v1/users/current/heartbeats.bulk\": EOF","now":"2022-08-27T10:39:52-03:00","version":"v1.54.0"}
{"caller":"pkg/api/api.go:64","func":"api.(*Client)","level":"warning","message":"dns error: https://api.wakatime.com/api/v1/users/current/statusbar/today. Retrying with fallback dns resolver","now":"2022-08-27T10:40:01-03:00","version":"v1.54.0"}
{"caller":"pkg/api/api.go:64","func":"api.(*Client)","level":"warning","message":"dns error: https://api.wakatime.com/api/v1/users/current/heartbeats.bulk. Retrying with fallback dns resolver","now":"2022-08-27T10:40:10-03:00","version":"v1.54.0"}
{"caller":"cmd/run.go:270","func":"cmd.runCmd","level":"error","message":"failed to run command: today fetch failed: failed fetching today from api: failed to make request to \"https://api.wakatime.com/api/v1/users/current/statusbar/today\": Get \"https://api.wakatime.com/api/v1/users/current/statusbar/today\": EOF","now":"2022-08-27T10:40:11-03:00","version":"v1.54.0"}
{"caller":"cmd/run.go:270","func":"cmd.runCmd","level":"error","message":"failed to run command: sending heartbeat(s) later due to api error: failed making request to \"https://api.wakatime.com/api/v1/users/current/heartbeats.bulk\": Post \"https://api.wakatime.com/api/v1/users/current/heartbeats.bulk\": EOF","now":"2022-08-27T10:40:20-03:00","version":"v1.54.0"}
{"caller":"pkg/api/api.go:64","func":"api.(*Client)","level":"warning","message":"dns error: https://api.wakatime.com/api/v1/users/current/heartbeats.bulk. Retrying with fallback dns resolver","now":"2022-08-27T10:40:52-03:00","version":"v1.54.0"}
{"caller":"cmd/run.go:270","func":"cmd.runCmd","level":"error","message":"failed to run command: sending heartbeat(s) later due to api error: failed making request to \"https://api.wakatime.com/api/v1/users/current/heartbeats.bulk\": Post \"https://api.wakatime.com/api/v1/users/current/heartbeats.bulk\": EOF","now":"2022-08-27T10:41:02-03:00","version":"v1.54.0"}
{"caller":"cmd/run.go:270","func":"cmd.runCmd","level":"error","message":"failed to run command: sending heartbeat(s) later because currently rate limited: won't send heartbeat due to backoff","now":"2022-08-27T10:42:00-03:00","version":"v1.54.0"}
{"caller":"pkg/api/api.go:64","func":"api.(*Client)","level":"warning","message":"dns error: https://api.wakatime.com/api/v1/users/current/statusbar/today. Retrying with fallback dns resolver","now":"2022-08-27T10:42:01-03:00","version":"v1.54.0"}
{"caller":"cmd/run.go:270","func":"cmd.runCmd","level":"error","message":"failed to run command: today fetch failed: failed fetching today from api: failed to make request to \"https://api.wakatime.com/api/v1/users/current/statusbar/today\": Get \"https://api.wakatime.com/api/v1/users/current/statusbar/today\": EOF","now":"2022-08-27T10:42:11-03:00","version":"v1.54.0"}
{"caller":"pkg/api/api.go:64","func":"api.(*Client)","level":"warning","message":"dns error: https://api.wakatime.com/api/v1/users/current/heartbeats.bulk. Retrying with fallback dns resolver","now":"2022-08-27T10:44:01-03:00","version":"v1.54.0"}
{"caller":"cmd/run.go:270","func":"cmd.runCmd","level":"error","message":"failed to run command: sending heartbeat(s) later due to api error: failed making request to \"https://api.wakatime.com/api/v1/users/current/heartbeats.bulk\": Post \"https://api.wakatime.com/api/v1/users/current/heartbeats.bulk\": EOF","now":"2022-08-27T10:44:11-03:00","version":"v1.54.0"}
```

How can I find these errors related to heartbeats logs that you mentioned?

**Comment by alanhamlett** [comment]
That's the right error message... It says it's a DNS error. Did something with your network change recently?

**Comment by andrefangeloni** [comment]
> That's the right error message... It says it's a DNS error. Did something with your network change recently?

No, I didn't. I'm using the same machine and the same internet provider at home.

**Comment by andrefangeloni** [comment]
@alanhamlett, I deleted the .wakatime.log folder and for testing, I'm using my mobile data internet and this is the log provided:

```json
{"caller":"cmd/run.go:270","func":"cmd.runCmd","level":"error","message":"failed to run command: sending heartbeat(s) later because currently rate limited: won't send heartbeat due to backoff","now":"2022-08-28T12:31:19-03:00","version":"v1.54.0"}
```

**Comment by alanhamlett** [comment]
`won't send heartbeat due to backoff` means previously there was a connection error so it's waiting a while.

You can reset the rate limit backoff by deleting `~/.wakatime-internal.cfg`. What error message do you see after deleting the internal file and using your mobile hotspot?

**Comment by andrefangeloni** [comment]
@alanhamlett I deleted the wakatime-internal.cfg file and with my mobile hotspot the data was synced, but today I worked 2:30h and VS Code is showing 44 min. this is the time I worked at the morning. But with my wi-fi internet the error still persists.

PS: I already restarted my modem, computer, VS Code...
When I click the link provided on logs, I got the the Unauthorized message on browser.
I've generated a new API KEY too, but seems the error isn't related with that.

Thanks in advance

```json
{"caller":"cmd/run.go:270","func":"cmd.runCmd","level":"error","message":"failed to run command: sending heartbeat(s) later due to api error: failed making request to \"https://api.wakatime.com/api/v1/users/current/heartbeats.bulk\": Post \"https://api.wakatime.com/api/v1/users/current/heartbeats.bulk\": EOF","now":"2022-08-28T21:09:23-03:00","version":"v1.54.0"}
{"caller":"pkg/api/api.go:64","func":"api.(*Client)","level":"warning","message":"dns error: https://api.wakatime.com/api/v1/users/current/statusbar/today. Retrying with fallback dns resolver","now":"2022-08-28T21:09:45-03:00","version":"v1.54.0"}
{"caller":"cmd/run.go:270","func":"cmd.runCmd","level":"error","message":"failed to run command: today fetch failed: failed fetching today from api: failed to make request to \"https://api.wakatime.com/api/v1/users/current/statusbar/today\": Get \"https://api.wakatime.com/api/v1/users/current/statusbar/today\": EOF","now":"2022-08-28T21:09:55-03:00","version":"v1.54.0"}
```

**Comment by andrefangeloni** [comment]
Hi, there.
I'm here to give you guys a feedback that the error still persists. 
I only can sync data with my mobile hotspot and even that, the behavior is a kind strange, because the timer on statusbar is different from the web dashboard.

I really like to use Wakatime, I'm using this extension more than a year and always worked like a charm.
If you guys need more information for debug, I'm here to help. Thanks

**Comment by eugene2k** [comment]
I have the same problem, and have had it for a while now (3 weeks or longer - I've deleted the log file, so can't say when it started). I'm on ArchLinux (amd64). Curiously, the problem goes away if wakatime is compiled without CGO_ENABLED=0.

**Comment by gandarez** [comment]
@eugene2k all binaries are already compiled without CGO enabled as you can see [here](https://github.com/wakatime/wakatime-cli/blob/develop/Makefile#L111).

**Comment by alanhamlett** [comment]
@andrefangeloni can you please extract [this past release](https://github.com/wakatime/wakatime-cli/releases/tag/v1.53.4) to `~/.wakatime/` and make sure the symlink `~/.wakatime/wakatime-cli` points to your newly extracted binary, then check if it works?

That will make sure it's not broken due to the Go 1.19 upgrade.

**Comment by eugene2k** [comment]
@gandarez, I said "without CGO_ENABLED=0", meaning "with CGO enabled". Sorry for the confusion.

**Comment by alanhamlett** [comment]
@eugene2k can you please extract [this past release](https://github.com/wakatime/wakatime-cli/releases/tag/v1.53.4) to `~/.wakatime/` and make sure the symlink `~/.wakatime/wakatime-cli` points to your newly extracted binary, then check if it works?

I want to make sure this isn't a bug caused by upgrading Go to 1.19.

**Comment by alanhamlett** [comment]
Might be fixed with #771, just released in [v1.55.1](https://github.com/wakatime/wakatime-cli/releases/tag/v1.55.1).

**Comment by eugene2k** [comment]
@alanhamlett, 
1.53.4 doesn't work. Same problem. 
1.55.1 does indeed work after reporting the DNS problem, 
the CGO build doesn't need to fall back to host ip, though - it just works

**Comment by alanhamlett** [comment]
Awesome, thanks for the detailed info! Now we're sure it's not related to Go 1.19, and most likely a pre-existing bug or limitation in Go's DNS implementation without CGO.

At least the latest version can send stats now, and we'll continue trying to get it working without the fallback.

**Comment by andrefangeloni** [comment]
Hi there, sorry for late response. But now it's working like a charm in my machine. I changed the default DNS from my internet provider to Cloudfare  (1.1.1.1). I compared with my co-workers and who are using another internet provider still working. Maybe something related with default DNS of my Internet. Thanks for support!

PS: After updated to the latest version (19.1.0), sync has stopped again =/
But this time there is no log to see what's happening

PS2: I revert my DNS to default from my internet provider and there are no errors, but sync isn't working and there are no logs showing too

**Comment by alanhamlett** [comment]
> PS: After updated to the latest version (19.1.0), sync has stopped again =/
> But this time there is no log to see what's happening

That's actually a different bug in the vscode plugin, which we fixed in vscode-wakatime release v19.2.0.

Upgrading your plugin in vscode will get it working again.
